File: bench/node/ackermann.js
// Ackermann ack(3,9) summed N times. Deep double-recursion (depth ~4093).
// Checksum = N * ack(3,9) = N * 4093.
const N = parseInt(process.env.BENCH_N || "6", 10);

function ack(m, k) {
  if (m === 0) return k + 1;
  if (k === 0) return ack(m - 1, 1);
  return ack(m - 1, ack(m, k - 1));
}

let total = 0;
for (let i = 0; i < N; i++) total += ack(3, 9);
console.log(total);

File: bench/ruby/ackermann.rb
# Ackermann ack(3,9) summed N times. Deep double-recursion (depth ~4093).
# Checksum = N * ack(3,9) = N * 4093.
n = (ENV["BENCH_N"] || "6").to_i

def ack(m, k)
  return k + 1 if m == 0
  return ack(m - 1, 1) if k == 0
  ack(m - 1, ack(m, k - 1))
end

total = 0
n.times { total += ack(3, 9) }
puts total
